tests/pytests/integration/utils/test_smb.py

```diff
@@ -0,0 +1,328 @@
+        username = getpass.getuser()
+        with salt.utils.files.fopen(passwdb, "w") as fp:
+            fp.write(
+                "{username}:0:XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX:AC8E657F8"
+                "3DF82BEEA5D43BDAF7800CC:[U          ]:LCT-507C14C7:"
+            )
+
+        samba_conf = Path(str(tmpdir) + os.sep + "smb.conf")
+        with salt.utils.files.fopen(samba_conf, "w") as fp:
+            fp.write(
+                f"[global]\n"
+                "realm = saltstack.com\n"
+                "interfaces = lo 127.0.0.0/8\n"
+                "smb ports = 1445\n"
+                "log level = 2\n"
+                "map to guest = Bad User\n"
+                "enable core files = no\n"
+                "passdb backend = smbpasswd\n"
+                "smb passwd file = {passwdb}\n"
+                "lock directory = {samba_dir}\n"
+                "state directory = {samba_dir}\n"
+                "cache directory = {samba_dir}\n"
+                "pid directory = {samba_dir}\n"
+                "private dir = {samba_dir}\n"
+                "ncalrpc dir = {samba_dir}\n"
+                "socket options = IPTOS_LOWDELAY TCP_NODELAY\n"
+                "min receivefile size = 0\n"
+                "write cache size = 0\n"
+                "client ntlmv2 auth = no\n"
+                "client min protocol = SMB3_11\n"
+                "client plaintext auth = no\n"
+                "\n"
+                "[public]\n"
+                "path = {public_dir}\n"
+                "read only = no\n"
+                "guest ok = no\n"
+                "writeable = yes\n"
+                "force user = {username}\n"
+            )
+        _smbd = subprocess.Popen([shutil.which("smbd"), "-F", "-P0", "-s", samba_conf])
+        time.sleep(1)
+
+        conn_dict = {
+            "tmpdir": tmpdir,
+            "samba_dir": samba_dir,
+            "public_dir": public_dir,
+            "passwdb": passwdb,
+            "username": username,
+            "samba_conf": samba_conf,
+        }
+
+        pidfile = Path(str(samba_dir) + os.sep + "smbd.pid")
+        assert pidfile.exists()
+        with salt.utils.files.fopen(pidfile, "r") as fp:
+            _pid = int(fp.read().strip())
+        if not check_pid(_pid):
+            raise Exception("Unable to locate smbd's pid file")
+
+        yield conn_dict
+
+        log.warning("teardown")
+        os.kill(_pid, signal.SIGTERM)
+
+
+def test_write_file_ipv4(smb_dict):
+    """
+    Transfer a file over SMB
+    """
+    name = "test_write_file_v4.txt"
+    content = "write test file content ipv4"
+    share_path = Path(str(smb_dict["public_dir"]) + os.sep + name)
+    assert not share_path.exists()
+
+    local_path = tempfile.mktemp()
+    with salt.utils.files.fopen(local_path, "w") as fp:
+        fp.write(content)
+    conn = salt.utils.smb.get_conn("127.0.0.1", smb_dict["username"], "foo", port=1445)
+    salt.utils.smb.put_file(local_path, name, "public", conn=conn)
+    conn.close()
```

Comment:
```suggestion
    with contextlib.closing(salt.utils.smb.get_conn("127.0.0.1", smb_dict["username"], "foo", port=1445)) as conn:
        salt.utils.smb.put_file(local_path, name, "public", conn=conn)
```

Let's do something like this for **all** of the `smb.get_conn`, in case of failures, it always get's closed, while the way you had it, if the `put_file` call was to raise an exception, the connection would be left open.

Reply:
yet another unittest issue port over I missed, I thoroughly disown the writer of the original test under unittest, naming no team developers :laughing: 
Will fix that.

Reply:
done